cdap-watchdog/src/main/java/io/cdap/cdap/metrics/collect/Distribution.java

```diff
@@ -0,0 +1,142 @@
+     */
+
+    /**
+     * ROWS is the number of rows in bucketCounts and COLS is the number of columns
+     * in bucketCounts. Note that ROWS*COLS MUST be less than or equal to 64 (Long.SIZE), since
+     * we are using a long bucketMask.
+     */
+    private static final int ROWS = 8;
+    private static final int COLS = 8;
+
+    private final long[][] bucketCounts = new long[ROWS][];
+
+    /**
+     * If count is zero then this field must be zero. Mean can be easily calculated from sum and count.
+     */
+    private double mean;
+
+    private long count;
+
+    private int numBuckets;
+
+    public void add(long value) {
+        int bucket = findBucketNumber(value);
+        // online calculation of mean and sumOfSquaredDeviation
+        count += 1;
+        double dev = value - mean;
+        mean += dev / count;
+
+        int i = getRow(bucket);
+        int j = getCol(bucket);
+        if (bucketCounts[i] == null) {
+            bucketCounts[i] = new long[COLS];
+        }
+        if (bucketCounts[i][j] == 0) {
+            numBuckets++;
+        }
+        bucketCounts[i][j] += 1;
+    }
+
+    private int findBucketNumber(long value) {
+        // negative values map to bucket 0
+        if (value < 0) {
+            return 0;
+        }
+        // value: 0 maps to bucket 1
+        if (value == 0) {
+            return 1;
+        }
+        // Long.numberOfTrailingZeros(Long.highestOneBit(value)) returns 0 for value: 1. Hence +2
+        return Math.min(Long.numberOfTrailingZeros(Long.highestOneBit(value)) + 2, Long.SIZE - 1);
+    }
+
+    private int getRow(int bucket) {
+        // Hotspot will optimize this division
+        // https://stackoverflow.com/questions/1514949/quick-java-optimization-question
+        return bucket / COLS;
+    }
+
+    private int getCol(int bucket) {
+        // Hotspot optimized this modulo
+        // https://stackoverflow.com/questions/29844345/how-is-the-modulo-operator-implemented-in-the-hotspot-jvm
+        return bucket % COLS;
+    }
+
+    public MetricValue getMetricValue(String metricName) {
+        long bucketMask = 0;
+        long[] bucketCountsArr = new long[numBuckets];
+        int bucketCountIndex = 0;
+        for (int row = 0; row < ROWS; row++) {
+            if (bucketCounts[row] == null) {
+                continue;
+            }
+            for (int col = 0; col < COLS; col++) {
+                if (bucketCounts[row][col] > 0) {
+                    bucketMask |= 1L << row * COLS + col;
+                    bucketCountsArr[bucketCountIndex] = bucketCounts[row][col];
+                    bucketCountIndex++;
+                }
+            }
+        }
+        return new MetricValue(metricName, bucketCountsArr, bucketMask, mean);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append(String.format("Distribution {count: %d, numBuckets: %d, mean:%32.2f %n",
```

Comment:
Don't add %n into toString. It's usually not expected where it's used. In most cases it will be just 1 or two rows printed